DOCS-630 Ansible docs (#270)
DataDog/ansible-datadog: README.md

# Ansible Datadog Role

The Ansible Datadog role installs and configures the Datadog Agent and integrations. Version `4` of the role installs the Datadog Agent v7 by default.

## Setup

### Requirements

- Requires Ansible v2.6+.
- Supports most Debian and RHEL-based Linux distributions, and Windows.

### Installation

Install the [Datadog role][1] from Ansible Galaxy on your Ansible server:

```shell
ansible-galaxy install datadog.datadog
```

To deploy the Datadog Agent on hosts, add the Datadog role and your API key to your playbook :

```text

> - hosts: servers
>   roles:
>     - { role: datadog.datadog, become: yes }
>     datadog_api_key: "<YOUR_DD_API_KEY>"

Reviewer: ```suggestion
- hosts: servers
  roles:
    - { role: datadog.datadog, become: yes }
  vars:
    datadog_api_key: "<YOUR_DD_API_KEY>"
```

# Ansible Datadog Role

The Ansible Datadog role installs and configures the Datadog Agent and integrations. Version `4` of the role installs the Datadog Agent v7 by default.

## Setup

### Requirements

- Requires Ansible v2.6+.
- Supports most Debian and RHEL-based Linux distributions, and Windows.

### Installation

Install the [Datadog role][1] from Ansible Galaxy on your Ansible server:

```shell
ansible-galaxy install datadog.datadog
```

> To deploy the Datadog Agent on hosts, add the Datadog role and your API key to your playbook :

Reviewer 2: ```suggestion
To deploy the Datadog Agent on hosts, add the Datadog role and your API key to your playbook:
```

| `datadog_config_ex`                        | (Optional) Extra INI sections to go in `/etc/dd-agent/datadog.conf` (Agent v5 only).                                                                                                                                                                                                                      |
| `datadog_apt_repo`                         | Override the default Datadog `apt` repository.                                                                                                                                                                                                                                                            |
| `datadog_apt_cache_valid_time`             | Override the default apt cache expiration time (defaults to 1 hour).                                                                                                                                                                                                                                      |

> | `datadog_apt_key_url_new`                  | Override the default URL to Datadog `apt` key (key ID `382E94DE` ; the deprecated `datadog_apt_key_url` variable refers to an expired key that's been removed from the role).                                                                                                                             |

Reviewer 2: ```suggestion
| `datadog_apt_key_url_new`                  | Override the default URL to Datadog `apt` key (key ID `382E94DE`; the deprecated `datadog_apt_key_url` variable refers to an expired key that's been removed from the role).                                                                                                                             |
```

| `datadog_additional_groups`                | Either a list, or a string containing a comma-separated list of additional groups for the `datadog_user` (Linux only).                                                                                                                                                                                    |
| `datadog_windows_ddagentuser_name`         | The name of Windows user to create/use, in the format `<domain>\<user>` (Windows only).                                                                                                                                                                                                                   |
| `datadog_windows_ddagentuser_password`     | The password used to create the user and/or register the service (Windows only).                                                                                                                                                                                                                          |

### Integrations

> To configure a Datadog integration (check), add an entry to the `datadog_checks` section. The first level key is the name of the check and the value is the YAML payload to write the configuration file. Examples are provided below.

Reviewer 2: ```suggestion
To configure a Datadog integration (check), add an entry to the `datadog_checks` section. The first level key is the name of the check, and the value is the YAML payload to write the configuration file. Examples are provided below.
```

          - some_data: true
```

#### Autodiscovery

When using Autodiscovery, there is no pre-processing nor post-processing on the YAML. This means every YAML section is added to the final configuration file, including `autodiscovery identifiers`.

> The example below configures the PostgeSQL check through **Autodiscovery**:

Reviewer 2: ```suggestion
The example below configures the PostgreSQL check through **Autodiscovery**:
```

            password: password
```

Learn more about [Autodiscovery in the Datadog documentation][3].

### Tracing

To enable trace collection with Agent v6 or v7 use the following configuration:

```yaml

> datadog_config:
>     apm_config:
>         enabled: true

Reviewer: Let's use the same indentation in all examples:
```suggestion
datadog_config:
  apm_config:
    enabled: true
```

```

To enable trace collection with Agent v5 use the following configuration:

```yaml

> datadog_config:
>     apm_enabled: "true" # has to be a string

Reviewer: ```suggestion
datadog_config:
  apm_enabled: "true" # has to be a string
```

For manual setup, refer to the [Network Performance Monitoring (NPM)][9] documentation.

#### Agent v5

To enable [live process][6] collection with Agent v5 use the following configuration:

```yml
datadog_config:
  process_agent_enabled: true
datadog_config_ex:
  process.config:
    scrub_args: true
    custom_sensitive_words: "<FIRST_WORD>,<SECOND_WORD>"
```

### Additional tasks

`pre_tasks` and `post_tasks` folders are available to run user defined tasks. `pre_tasks` run before executing any tasks from the Datadog Ansible role and `post_tasks` run after execution of the role.

Installation tasks on supported platforms register the variable `datadog_agent_install`, which can be used in `post_tasks` to check the installation task's result. `datadog_agent_install.changed` is set to `true` if the installation task did install something, and `false` otherwise (for instance if the requested version was already installed).

## Versions

By default, the current major version of the Datadog Ansible role installs Agent v7. The variables `datadog_agent_version` and `datadog_agent_major_version` are available to control the Agent version installed.

> For v3+ of this role, when `datadog_agent_version` is used to pin a specific Agent version, the role derives per-OS version names to comply with the version naming schemes of the supported operating systems, for example:

Reviewer: That was a typo, let's fix it:
```suggestion
For v4+ of this role, when `datadog_agent_version` is used to pin a specific Agent version, the role derives per-OS version names to comply with the version naming schemes of the supported operating systems, for example:
```

### Repositories

#### Linux

When the variables `datadog_apt_repo`, `datadog_yum_repo`, and `datadog_zypper_repo` are not set, the official Datadog repositories for the major version set in `datadog_agent_major_version` are used:

| # | Default apt repository                    | Default yum repository             | Default zypper repository               |
|---|-------------------------------------------|------------------------------------|-----------------------------------------|
| 5 | deb https://apt.datadoghq.com stable main | https://yum.datadoghq.com/rpm      | https://yum.datadoghq.com/suse/rpm      |
| 6 | deb https://apt.datadoghq.com stable 6    | https://yum.datadoghq.com/stable/6 | https://yum.datadoghq.com/suse/stable/6 |
| 7 | deb https://apt.datadoghq.com stable 7    | https://yum.datadoghq.com/stable/7 | https://yum.datadoghq.com/suse/stable/7 |

To override the default behavior, set these variables to something else than an empty string.

> If you previously used the Agent v5 variables, use the **New** variables below with `datadog_agent_major_version` set to `5` or `datadog_agent_version` pinned to a specific Agent v5 version.

Reviewer: ```suggestion
If you previously used the Agent v5 variables, use the **new** variables below with `datadog_agent_major_version` set to `5` or `datadog_agent_version` pinned to a specific Agent v5 version.
```

            password: password
```

Learn more about [Autodiscovery in the Datadog documentation][3].

### Tracing

To enable trace collection with Agent v6 or v7 use the following configuration:

```yaml
datadog_config:
  apm_config:
    enabled: true
```

To enable trace collection with Agent v5 use the following configuration:

```yaml
datadog_config:
  apm_enabled: "true" # has to be a string
```

### Live processes

To enable [live process][6] collection with Agent v6 or v7 use the following configuration:

```yml
datadog_config:
  process_config:
    enabled: "true" # type: string
```

The possible values for `enabled` are: `"true"`, `"false"` (only container collection), or `"disabled"` (disable live processes entirely).

#### Variables

The following variables are available for live processes:

* `scrub_args`: Enables the scrubbing of sensitive arguments from a process command line (defaults to `true`).

> * `custom_sensitive_words`: Expands the default list of sensitive words used by the cmdline scrubber.

Reviewer 2: ```suggestion
* `custom_sensitive_words`: Expands the default list of sensitive words used by the command line scrubber.
```

#### System Probe

> The [network performance monitoring][7] system probe is configured under the `system_probe_config` variable. Any variables nested underneath are written to the `system-probe.yaml`.

Reviewer 2: ```suggestion
The [Network Performance Monitoring][7] (NPM) system probe is configured under the `system_probe_config` variable. Any variables nested underneath are written to the `system-probe.yaml`.
```

**Note**: The system probe only works on Linux with the Agent v6+.

#### Example configuration

```yml
datadog_config:
  process_config:
    enabled: "true" # type: string
    scrub_args: true
    custom_sensitive_words: ['consul_token','dd_api_key']
system_probe_config:
  enabled: true
  sysprobe_socket: /opt/datadog-agent/run/sysprobe.sock
```

Once modification is complete, follow the steps below:

1. Start the system-probe: `sudo service datadog-agent-sysprobe start` **Note**: If the service wrapper is not available on your system, run this command instead: `sudo initctl start datadog-agent-sysprobe`.
2. [Restart the Agent][8]: `sudo service datadog-agent restart`.
3. Enable the system-probe to start on boot: `sudo service enable datadog-agent-sysprobe`.

> For manual setup, refer to the [Network Performance Monitoring (NPM)][9] documentation.

Reviewer 2: ```suggestion
For manual setup, refer to the [NPM][9] documentation.
```

#### Agent v5

> To enable [live process][6] collection with Agent v5 use the following configuration:

Reviewer 2: ```suggestion
To enable [live process][6] collection with Agent v5, use the following configuration:
```

```yml
datadog_config:
  process_agent_enabled: true
datadog_config_ex:
  process.config:
    scrub_args: true
    custom_sensitive_words: "<FIRST_WORD>,<SECOND_WORD>"
```

### Additional tasks

> `pre_tasks` and `post_tasks` folders are available to run user defined tasks. `pre_tasks` run before executing any tasks from the Datadog Ansible role and `post_tasks` run after execution of the role.

Reviewer 2: ```suggestion
`pre_tasks` and `post_tasks` folders are available to run user defined tasks. `pre_tasks` run before executing any tasks from the Datadog Ansible role, and `post_tasks` run after execution of the role.
```

|---|-------------------------------------------|------------------------------------|-----------------------------------------|
| 5 | deb https://apt.datadoghq.com stable main | https://yum.datadoghq.com/rpm      | https://yum.datadoghq.com/suse/rpm      |
| 6 | deb https://apt.datadoghq.com stable 6    | https://yum.datadoghq.com/stable/6 | https://yum.datadoghq.com/suse/stable/6 |
| 7 | deb https://apt.datadoghq.com stable 7    | https://yum.datadoghq.com/stable/7 | https://yum.datadoghq.com/suse/stable/7 |

To override the default behavior, set these variables to something else than an empty string.

If you previously used the Agent v5 variables, use the **new** variables below with `datadog_agent_major_version` set to `5` or `datadog_agent_version` pinned to a specific Agent v5 version.

| Old                          | New                   |
|------------------------------|-----------------------|
| `datadog_agent5_apt_repo`    | `datadog_apt_repo`    |
| `datadog_agent5_yum_repo`    | `datadog_yum_repo`    |
| `datadog_agent5_zypper_repo` | `datadog_zypper_repo` |

#### Windows

> When the variable `datadog_windows_download_url` is not set, the official Windows msi package corresponding to the `datadog_agent_major_version` is used:

Reviewer 2: ```suggestion
When the variable `datadog_windows_download_url` is not set, the official Windows MSI package corresponding to the `datadog_agent_major_version` is used:
```

> | # | Default Windows msi package URL                                                  |

Reviewer 2: ```suggestion
| # | Default Windows MSI package URL                                                  |
```

     action: remove
```

To see the available versions of Datadog integrations, refer to their `CHANGELOG.md` file in the [integrations-core repository][5].

### Downgrade

To downgrade to a prior version of the Agent:

1. Set `datadog_agent_version` to a specific version, for example: `5.32.5`.
2. Set `datadog_agent_allow_downgrade` to `yes`.

**Notes:**

> - Downgrades for Centos only work with Ansible 2.4+.

Reviewer: Probably not needed since the current version of the role requires Ansible 2.6+
```suggestion
```

    ...
```

**Note**: This configuration fails on Linux hosts. Only use it if the playbook is specific to Windows hosts. Otherwise, use the [inventory file method](#inventory-file).

## Troubleshooting

### Debian stretch

On Debian stretch, the `apt_key` module used by the role requires an additional system dependency to work correctly. The dependency (`dirmngr`) is not provided by the module. Add the following configuration to your playbooks to make use of the present role:

```yml
---
- hosts: all
  pre_tasks:
    - name: Debian Stretch requires the dirmngr package to use apt_key
      become: yes
      apt:
        name: dirmngr
        state: present

  roles:

>     - { role: datadog.datadog, become: yes, datadog_api_key: "<YOUR_DD_API_KEY>" }

Reviewer: Let's use the same playbook structure in all examples
```suggestion
    - { role: datadog.datadog, become: yes }
  vars:
    datadog_api_key: "<YOUR_DD_API_KEY>"
```

    ...
```

**Note**: This configuration fails on Linux hosts. Only use it if the playbook is specific to Windows hosts. Otherwise, use the [inventory file method](#inventory-file).

## Troubleshooting

### Debian stretch

> On Debian stretch, the `apt_key` module used by the role requires an additional system dependency to work correctly. The dependency (`dirmngr`) is not provided by the module. Add the following configuration to your playbooks to make use of the present role:

Reviewer 2: ```suggestion
On Debian Stretch, the `apt_key` module used by the role requires an additional system dependency to work correctly. The dependency (`dirmngr`) is not provided by the module. Add the following configuration to your playbooks to make use of the present role:
```
